flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCheckpointIDCounter.java

```diff
@@ -77,14 +84,26 @@ public ZooKeeperCheckpointIDCounter(CuratorFramework client, String counterPath)
 		this.client = checkNotNull(client, "Curator client");
 		this.counterPath = checkNotNull(counterPath, "Counter path");
 		this.sharedCount = new SharedCount(client, counterPath, 1);
+
+		this.connectionStateListeners = new ArrayList<>();
+		this.connectionStateListeners.add((ignore, newState) -> lastState = newState);
+	}
+
+	@VisibleForTesting
+	ZooKeeperCheckpointIDCounter(CuratorFramework client, String counterPath, Collection<ConnectionStateListener> listeners) {
+		this(client, counterPath);
+		this.connectionStateListeners.addAll(listeners);
 	}
 
 	@Override
 	public void start() throws Exception {
 		synchronized (startStopLock) {
 			if (!isStarted) {
 				sharedCount.start();
-				client.getConnectionStateListenable().addListener(connStateListener);
+
+				for (ConnectionStateListener listener : connectionStateListeners) {
+					client.getConnectionStateListenable().addListener(listener);
```

Comment:
Since `getConnectionStateListenable` does not guarantee the order in which the listener are called, the added test case is unstable (if the testing listener is called before the one which sets `lastState`). I suggest to introduce a `LastStateConnectionStateListener` (similar to `SharedCountConnectionStateListener`) which we pass into the class.

Reply:
Nice catch!

Reply:
Another way is that we instead implement a chain of listeners so that the order is deterministic.

Reply:
I also thought about this but I think the other approach is easier to understand.